eclass/python-r1.eclass

```diff
@@ -683,7 +683,7 @@ python_foreach_impl() {
 #	$(python_gen_any_dep 'dev-python/epydoc[${PYTHON_USEDEP}]' 'python2*') )"
 #
 # python_check_deps() {
-#	has_version "dev-python/epydoc[${PYTHON_USEDEP}]"
+# 	use !doc || python_has_version "dev-python/epydoc[${PYTHON_USEDEP}]"
```

Comment:
I think the newer examples prefer:

```
use doc || return 0
python_has_version ...
```

Reply:
Okay, I'll change it to that for consistency. It's functionally identical, though, right? And what if you had multiple different sets of use flags to check?
```
BDEPEND="$(python_gen_any_dep '
    foo? ( dev-python/foo[${PYTHON_USEDEP}] )
    bar? ( dev-python/bar[${PYTHON_USEDEP}] )
')"
```
Then what would you put in `python_check_deps()`? I like my way because it structurally mirrors the syntax of `BDEPEND`.
```
python_check_deps() {
    { use !foo || python_has_version "dev-python/foo[${PYTHON_USEDEP}]"; } &&
    { use !bar || python_has_version "dev-python/bar[${PYTHON_USEDEP}]"; }
}
```

Reply:
But that thing is horribly hard to read. Even the `foo && { bar || baz; }` thing is ugly. To be honest, I'm wondering if maybe we should just remove the examples from eclassdoc and expect people to read the Guide.

Reply:
> But that thing is horribly hard to read.

Is it? Maybe for a complete n00b of a programmer. But again, I really was asking what you would put in `python_check_deps()` in such a case. Would you write it all out like this?:
```
python_check_deps() {
    if use foo ; then
        python_has_version "dev-python/foo[${PYTHON_USEDEP}]" || return $?
    fi
    if use bar ; then
        python_has_version "dev-python/bar[${PYTHON_USEDEP}]" || return $?
    fi
}
```
Are you really arguing that that's less "horrible" than the compact form that I wrote using `&&`?

Anyway, I find the whole mechanism of `python_check_deps()` really awkward, as it requires duplicating the `python_gen_any_dep` dependencies of the package but with different syntax. It would be so much nicer if an ebuild author could write:
```
BDEPEND="$(python_gen_any_dep '
    dev-python/foo[${PYTHON_SINGLE_USEDEP}]
    || ( dev-python/bar[${PYTHON_USEDEP}]
         dev-python/baz[${PYTHON_USEDEP}] )
')"

python_get_any_deps() {
    echo "
        dev-python/foo[${PYTHON_SINGLE_USEDEP}]
        || ( dev-python/bar[${PYTHON_USEDEP}]
             dev-python/baz[${PYTHON_USEDEP}] )
    "
}
```
And then the eclass would handle parsing of the returned dependency spec, including the `|| ( … )` syntax. This would allow a straight copy+paste of the value of the `python_gen_any_dep` argument to the `python_get_any_deps()` output, with just a change from single-quoting to double-quoting to enable variable substitution.

This solution could co-exist with `python_check_deps()` simply by having the eclass define a default `python_check_deps()` that calls `python_get_any_deps` and checks the returned dependency spec if `python_get_any_deps` is defined or else does nothing if it's not defined. Existing ebuilds would be overriding the default `python_check_deps()` and thus would continue to work exactly as they do now.

Reply:
I have implemented the `python_get_any_deps()` alternative in #24677. Please have a look there.